addons/binding/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/VerisureSession.java

```diff
@@ -0,0 +1,513 @@
+import org.openhab.binding.verisure.internal.model.VerisureSmartPlugJSON;
+import org.openhab.binding.verisure.internal.model.VerisureThingJSON;
+import org.openhab.binding.verisure.internal.model.VerisureUserPresenceJSON;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * This class performs the communication with Verisure My Pages.
+ *
+ * @author Jarle Hjortland - Initial contribution
+ * @author Jan Gustafsson - Re-design and support for several sites
+ *
+ */
+@NonNullByDefault
+public class VerisureSession {
+    private final class VerisureInstallation {
+        private @Nullable String installationName;
+        private int installationInstance;
+        private @Nullable BigDecimal installationId;
+
+        public VerisureInstallation(@Nullable String installationName) {
+            this.installationName = installationName;
+        }
+
+        public @Nullable BigDecimal getInstallationId() {
+            return installationId;
+        }
+
+        public int getInstallationInstance() {
+            return installationInstance;
+        }
+
+        public @Nullable String getInstallationName() {
+            return installationName;
+        }
+
+        public void setInstallationId(BigDecimal installationId) {
+            this.installationId = installationId;
+        }
+
+        public void setInstallationInstance(int installationInstance) {
+            this.installationInstance = installationInstance;
+        }
+    }
+
+    private final HashMap<String, org.openhab.binding.verisure.internal.model.VerisureThingJSON> verisureThings = new HashMap<String, org.openhab.binding.verisure.internal.model.VerisureThingJSON>();
+    private final Logger logger = LoggerFactory.getLogger(VerisureSession.class);
+    private final Gson gson = new GsonBuilder().create();
+    private final List<DeviceStatusListener> deviceStatusListeners = new CopyOnWriteArrayList<>();
+    private final Hashtable<String, @Nullable VerisureInstallation> verisureInstallations = new Hashtable<String, @Nullable VerisureInstallation>();
+
+    private boolean areWeLoggedOut = true;
+    private @Nullable String authstring;
+    private @Nullable String csrf;
+    private @Nullable BigDecimal pinCode;
+    private @Nullable BigDecimal numberOfInstallations;
+    private HttpClient httpClient;
+
+    public VerisureSession(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public void initialize(@Nullable String authstring, @Nullable BigDecimal pinCode,
+            @Nullable BigDecimal numberOfInstallations) {
+        logger.debug("VerisureSession:initialize");
+        if (authstring != null) {
+            this.authstring = authstring.substring(0);
+            this.pinCode = pinCode;
+            this.numberOfInstallations = numberOfInstallations;
+            // Try to login to Verisure
+            if (logIn()) {
```

Comment:
You should do this in background as this could take some time. OpenHAB framework waits 5000ms to initialization be complete. Additionally this could fail (e.g. if internet is down), so biding should do retries. BaseThingHandler provides your ScheduledExecutorService which can be used.